Alignment configuration free end gaps

smehringer opened the pull request:
Resolves seqan/product_backlog/issues/96

The alignment_configurator adaption is probably not the smartest move but it is part of the old alignment and will be deleted anyway.

marehr commented on include/seqan3/alignment/pairwise/detail/policy_alignment_matrix.hpp:
```diff
@@ -100,23 +100,20 @@ class policy_alignment_matrix
-        {
-            invalid_band |= (upper_diagonal < 0 || lower_diagonal > 0);
-            error_cause += " The first cell of the matrix is not enclosed by the band.";
-        }
```
?

marehr commented on include/seqan3/alignment/pairwise/detail/policy_alignment_matrix.hpp:
```diff
@@ -100,23 +100,20 @@ class policy_alignment_matrix
 
         bool invalid_band = upper_diagonal < lower_diagonal;
         std::string error_cause = (invalid_band) ? " The upper diagonal is smaller than the lower diagonal." : "";
-        if constexpr (traits_t::with_free_end_gaps)
```
?

marehr commented:
Is every single commit "buildable"? If so could we split this PR into multiple ones, this would make it easier on the reviewers (I would review all of them, because I know somehow now what this big PR is about).

smehringer commented:
> 1st_place_medal For this very Ren[n]e like PR safety_vest construction_worker_man

I did my best.

> This is just the first round, I have some problem to find the important changes. Even some https://www.youtube.com/watch?v=DU2bgWOln64 couldn't help me xD
> I'll go commit-by-commit again, tomorrow.
> 
> Is this a follow-up PR of other changes that I'm maybe unaware of?

Yes sorry, it is the second part of seqan/product_backlog/issues/96, following PR https://github.com/seqan/seqan3/pull/1938

> Is every single commit "buildable"? 

Yes

> If so could we split this PR into multiple ones, this would make it easier on the reviewers (I would review all of them, because I now know somehow what this big PR is about).

I know this will sound childish: Rene never has to :(
And actually the commits do belong together quite strongly. In the first part in PR https://github.com/seqan/seqan3/pull/1938
I just introduced the `method_global` configuration without using it and without initializing its free end gap members.
In this part I do (by commit):

- Add the free and gap initialisation wherever it is needed in a `method_global` configuration
- Adapt the "new alignment code" to actually use `method_global` instead of `aligned_ends`
- Adapt the "old alignment code" to actually use `method_global` instead of `aligned_ends`
- Delete all now unnecessary usages of the `aligned_ends` config
- Documentation
- Moving `aligned_ends` to `detail::aligned_ends` because it is still needed in the "old alignment code" (this probably fits better after adapt the old alignment code.. I can change that)

So I already tried to split this in commits although it is basically the single task of "replace aligned_ends with method_global".
I thought in single PRs the changes won't have much meaning but I can make single PR's if you like.

smehringer commented on include/seqan3/alignment/pairwise/alignment_configurator.hpp:
```diff
@@ -319,14 +320,16 @@ struct alignment_configurator
+                    !(method_global_cfg.free_end_gaps_sequence2_leading.get() ||
+                      method_global_cfg.free_end_gaps_sequence2_trailing.get()) &&
+                    (method_global_cfg.free_end_gaps_sequence1_leading.get() ==
+                     method_global_cfg.free_end_gaps_sequence1_trailing.get()))
```
Oh you are right. Um, I just assumed to also use strong types but I have no strong feelings. Shall I change them to boolean?

marehr replied:
Yes, this should be changed.

smehringer commented:
superseeded by separate PRs
